catalog: add SystemDatabaseSchemaVersion field to DatabaseDescriptor

andyyang890 opened the pull request:
**catalog: add SystemDatabaseSchemaVersion field to DatabaseDescriptor**

This patch adds a new `SystemDatabaseSchemaVersion` field to
`descpb.DatabaseDescriptor` that will be used to store the schema
version of the system database.

Release note: None

---

**upgrades: add helper function to bump SystemDatabaseSchemaVersion**

This patch adds a helper function that should be used by upgrades that
create or modify the schema of system tables to bump the schema version
field on the `system` database descriptor.

Release note: None

---

Fixes #109039

andyyang890 committed:
catalog: add SystemDatabaseSchemaVersion field to DatabaseDescriptor

This patch adds a new `SystemDatabaseSchemaVersion` field to
`descpb.DatabaseDescriptor` that will be used to store the schema
version of the system database.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111404)
<!-- Reviewable:end -->

andyyang890 committed:
upgrades: add helper function to bump SystemDatabaseSchemaVersion

This patch adds a helper function that should be used by upgrades that
create or modify the schema of system tables to bump the schema version
field on the `system` database descriptor.

Release note: None

dt commented on pkg/upgrade/upgrades/schema_changes.go:
```diff
@@ -411,3 +412,28 @@ func onlyHasColumnFamily(
+func bumpSystemDatabaseSchemaVersion(
+	ctx context.Context, cs clusterversion.ClusterVersion, d upgrade.TenantDeps,
+) error {
+	return d.DB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) error {
```
should this include a WaitForOneVersion, so that we know anyone who observes the cluster version changed -- which will be done after the upgrade func this is called in returns -- won't also still have a lease of the old db desc?

andyyang890 replied:
This is already implicitly done in `DescsTxn`!

https://github.com/cockroachdb/cockroach/blob/173a6db0eacdcc0de72fd2b8542a9c6662deda5e/pkg/sql/internal.go#L1668
https://github.com/cockroachdb/cockroach/blob/173a6db0eacdcc0de72fd2b8542a9c6662deda5e/pkg/sql/internal.go#L1708

dt replied:
Oh, fancy! I've been out of the loop for awhile here I guess.

dt commented on pkg/sql/catalog/systemschema/system.go:
```diff
@@ -1056,13 +1057,15 @@ const SystemDatabaseName = catconstants.SystemDatabaseName
 // descriptor.
 func MakeSystemDatabaseDesc() catalog.DatabaseDescriptor {
 	priv := privilege.List{privilege.CONNECT}
+	binaryVersion := clusterversion.ByKey(clusterversion.BinaryVersionKey)
```
I think init'ing this to binary version will mean we can't bump it in any permanent migrations, and is maybe a problem for when we bootstrap to the old schema that was as of minBinary, e.g. during tenant creation. I think we want to plumb in the system schema version that we're bootstrapping right?

dt replied:
I think another thing we might want to do is make a var somewhere -- maybe in systemschema -- that anyone who adds a bumping migration then bumps, and use that, rather than clusterversion.BinaryVersionKey, for the default value during bootstrap. This will mean one more step for migration-that-bumps-it authors, but importantly means folks who add a CV _that doesn't bump the system schema version_ don't need to regenerate all the testdata / catalog dumps / etc?

fqazi reviewed (changes requested):
Just some minor things. Great work, Andy!

> Reviewed 5 of 10 files at r1, 12 of 12 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111404#-:-Nfq5weaBEkvWQzdLm7X:bmyheqw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @knz, and @rafiss)

___
*[`pkg/sql/catalog/systemschema/system.go` line 1059 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111404#-Nfq5G0V5rSnMclYKISu:-Nfq5G0V5rSnMclYKISv:b-871eh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b0ad5cd874e8776e279c4497a06dd0b6cb13e08/pkg/sql/catalog/systemschema/system.go#L1059)):*
> ```Go
> // that should be used during bootstrap. It should be bumped up alongside any
> // upgrade that creates or modifies the schema of a system table.
> var SystemDatabaseSchemaBootstrapVersion = clusterversion.ByKey(clusterversion.V23_2_RegionaLivenessTable)
> ```

Should we add something to the bootstrap test so that we make sure people bump this up?
___
*[`pkg/upgrade/upgrades/schema_changes.go` line 428 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111404#-Nfq5i2v5N9XTCKyc2ie:-Nfq5i2v5N9XTCKyc2if:b-ghq3f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d793e6712596067875a017fa09828111c8369e7f/pkg/upgrade/upgrades/schema_changes.go#L428)):*
> ```Go
> 		}
> 		if sv := systemDBDesc.GetSystemDatabaseSchemaVersion(); sv != nil {
> 			if cs.Version.Less(*sv) {
> ```

I think the validation layer will also catch it right before we write. So this might be redundant?


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111404)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @knz, and @rafiss)

___
*[`pkg/sql/catalog/systemschema/system.go` line 1059 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111404#-Nfq5G0V5rSnMclYKISu:-NfreExP2wrXTr7Z03n_:bl3bei8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b0ad5cd874e8776e279c4497a06dd0b6cb13e08/pkg/sql/catalog/systemschema/system.go#L1059)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Should we add something to the bootstrap test so that we make sure people bump this up?
</blockquote></details>

Added a new test to check this!
___
*[`pkg/upgrade/upgrades/schema_changes.go` line 428 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111404#-Nfq5i2v5N9XTCKyc2ie:-NfreKOY-tXKnBBb_QYY:b-r6edzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d793e6712596067875a017fa09828111c8369e7f/pkg/upgrade/upgrades/schema_changes.go#L428)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think the validation layer will also catch it right before we write. So this might be redundant?
</blockquote></details>

We discussed this offline already but just noting it here for posterity, the check in `ValidateSelf` is checking that we're setting it to a valid version relative to `BinaryMinSupportedVersion` and `BinaryVersion` while this check is making sure we can't decrease the version.

We also discussed moving that check to `ValidateSelf` but since that operates only on an `*dbdesc.immutable` (not an `*dbdesc.Mutable` which has access to the previous descriptor), we decided that was too complex and unnecessary for now.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r5, 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111404#-:-Nfs4zfMD_SnDzR4dkt-:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andyyang890, @knz, and @rafiss)

___
*[`pkg/sql/catalog/systemschema/system.go` line 1059 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111404#-Nfq5G0V5rSnMclYKISu:-Nfs559T3HPf026rrAy1:bpcrwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b0ad5cd874e8776e279c4497a06dd0b6cb13e08/pkg/sql/catalog/systemschema/system.go#L1059)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Added a new test to check this!
</blockquote></details>

Great work 😊


<!-- Sent from Reviewable.io -->

andyyang890 commented:
TFTRs!

bors r=dt,fqazi
